helix-core/src/main/java/org/apache/helix/controller/rebalancer/strategy/AbstractEvenDistributionRebalanceStrategy.java

```diff
@@ -82,7 +85,23 @@ public void init(String resourceName, final List<String> partitions,
   @Override
   public ZNRecord computePartitionAssignment(final List<String> allNodes,
       final List<String> liveNodes, final Map<String, Map<String, String>> currentMapping,
-      ResourceControllerDataProvider clusterData) throws HelixException {
+      ResourceControllerDataProvider clusterData) {
+    // only compute assignments for instances with available spaces
+    return computeBestPartitionAssignment(
+        getNonEmptyWeightNodes(allNodes, clusterData.getInstanceConfigMap()),
+        getNonEmptyWeightNodes(liveNodes, clusterData.getInstanceConfigMap()), currentMapping,
+        clusterData);
+  }
+
+  private List<String> getNonEmptyWeightNodes(List<String> nodes, Map<String, InstanceConfig> instanceConfigMap) {
+    if (instanceConfigMap == null) {
+      return nodes;
+    }
+    return nodes.stream().filter(node -> instanceConfigMap.get(node).getWeight() != 0).collect(Collectors.toList());
```

Comment:
What if instance config map does not contain the instance name? This filter will throw NPE.

Reply:
Added the check, it will throw HelixException if no instance config found.

Reply:
Note that there is no guarantee the all node list will contain all live nodes. So what you checked will not pretect liveNodes. In my opinion, we just check and throw Exception in getNonZeroWeightNodes.

Reply:
I've noticed the strange behavior in local testing. By reading the code, allNodes means all the instances. They should include the liveNodes as subset. I don't know which part of helix is messing up the things. If it errors out due to the exception, it is also true that some resource of weight = 0 instance could stick there why the rebalancer failed to compute the new ideal mapping.

Reply:
allNodes means all the nodes with instance config. There is no guarantee a live instance has the config. When rebalance, just ignore the live instance with no config.

Let's just check for allNodes. That shall be good enough.

Reply:
Updated the PR, and the checks are only for `allNodes` only. I've also manually tested the changes on some cluster, setting one instance weight = 0, the result is expected.